Split calls to ocamldep

diml opened the pull request:
Instead of doing a single call to ocamldep, do one per file. This will allow to support `menhir --infer` (#305).

rgrinberg commented:
The cram tests need to be updated.

Curious about why is this change desired. Doesn't this slow builds by invoking an external command potentially a lot more often?

rgrinberg commented on src/module_compilation.ml:
```diff
@@ -36,20 +36,15 @@ let build_cm sctx ?sandbox ~dynlink ~flags ~cm_kind ~(dep_graph:Ocamldep.dep_gra
-                 | Cmi | Cmo -> [Module.cm_file m ~dir Cmi]
-                 | Cmx -> [Module.cm_file m ~dir Cmi; Module.cm_file m ~dir Cmx])))
+          (Ocamldep.Dep_graph.deps_of dep_graph m >>^ fun deps ->
+           List.concat_map     deps
```
An accidental space snuck in here

diml replied:
oops

diml commented:
I'll check the performance implications. The motivations are:

- it's the only I can think of to support `menhir --infer`, i.e. we need to compile some .cmi files before we can generate some other .ml files
- it might improve parallelism in some cases, since we'll be able to start building some modules before all the .ml are generated
- it seems that sometimes users like to be able to start building with .ml files that are not finished and have syntax errors, this will allow it

diml commented:
I tried to build jenga from scratch with -j1 and -j20. It is slightly slower indeed:

```
master -j1

real	2m58.795s
user	2m16.904s
sys	0m25.476s

split -j1

real	3m7.211s
user	2m18.196s
sys	0m30.108s

master -j20

real	0m49.544s
user	3m36.376s
sys	0m33.088s

split -j20

real	0m52.668s
user	3m32.132s
sys	0m36.928s
```

Though I'm still in favour of this patch since it allows to support `menhir --infer`. Additionally, if we implement something like https://github.com/ocaml/dune/issues/193#issuecomment-359240107, then this won't matter anyway and the new API is more friendly to such a change.

diml commented on src/ordered_set_lang.ml:
```diff
@@ -120,10 +120,8 @@ module Unexpanded = struct
-        String_map.find_exn (f fn) files_contents ~string_of_key:(sprintf "%S")
-          ~desc:(fun _ -> "<filename to s-expression>")
-      ) ~f:(fun s -> f (String_with_vars.t s))
+      parse_general (Option.value_exn (String_map.find (f fn) files_contents))
```
This code is only called from `Super_context.expand_and_eval_set` where it's easy to check the correctness. I hesitated between doing this and upgrading this code to `Sexp.code_error` but maybe I should, just for uniformity

diml replied:
done

diml commented:
I'm merging this as the slowdown is relatively small, we have a way to improve things and I need this PR for another one I'm preparing.

diml commented:
Well, I'll merge it once node is installed and I can re-generate the js_of_ocaml tests...

rgrinberg commented:
Ok. Fine with me. Slow doesnt seem so bad and it might perform better on
some incremental builds

On Tue, Feb 6, 2018, 7:46 PM Jérémie Dimino <notifications@github.com>
wrote:

> Well, I'll merge it once node is installed and I can re-generate the
> js_of_ocaml tests...
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/ocaml/dune/pull/486#issuecomment-363398513>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAIe-2d1pDPUFGgg9QCE87vKvbsC6q-Yks5tSDt_gaJpZM4R5-HX>
> .
>
